A15-binary-logging

```diff
@@ -0,0 +1,264 @@
+# gRPC Binary Logging
+
+*   Author(s): Yuchen Zeng
+*   Approver: a11r
+*   Status: In Review
+*   Implemented in:
+*   Last updated: 10/11/17
+*   Discussion at:
+
+## Abstract
+
+Provide binary logging functionality in gRPC.
+
+## Background
+
+gRPC binary logging logs RPCs in binary format. RPC logs generated from
+production traffic will be useful in the following cases:
+
+-   Troubleshooting services, find exceptions
+-   Loadtesting
+-   Reply RPCs from production
+
+This functionality should be added in each of our supported languages. The ease
+of adoption should also be considered.
+
+## Proposal
+
+The proposed design is adding binary logging as a built-in functionality in
+gRPC. Users can turn it on/off by setting the `GRPC_BINARY_LOG_FILTER`
+environment variable. It's not required to modify application code to generate
+logs. The control interface should allow users to change the amount and contents
+of logs.
+
+### Log Format
+
+gRPC logs each RPC in the format defined by proto GrpcLogEntry. Initial/trailing
+metadata and messages are logged separately to match the on-the-wire
+representation.
+
+The proto of GrpcLogEntry:
+
+```
+// Log entry we store in binary logs
+message GrpcLogEntry {
+  // Enumerates the type of logs
+  enum Type {
+    SEND_INITIAL_METADATA = 0;
```

Comment:
This enum should ideally have an `UNKNOWN=0` to make it future-proof.  Or the enum could be removed in lieu of the fact that we're using a `oneof` below.

Reply:
The current proto has an `UNKOWN`